src/operation/recursive.rs

```diff
@@ -242,31 +250,83 @@ impl Op {
                         }
                     }
                 }
+
+                let (from_file, metadata, mut to_file) = futures::try_join!(
+                    async {
+                        compio::fs::OpenOptions::new()
+                            .read(true)
+                            .open(&self.from)
+                            .await
+                    },
+                    compio::fs::metadata(&self.from),
+                    // This is atomic and ensures `to` is not created by any other process
+                    async {
+                        compio::fs::OpenOptions::new()
+                            .create_new(true)
+                            .write(true)
+                            .open(&self.to)
+                            .await
+                    }
+                )?;
+
                 progress.total_bytes = Some(metadata.len());
                 (ctx.on_progress)(self, &progress);
-                // This is atomic and ensures `to` is not created by any other process
-                let mut to_file = fs::OpenOptions::new()
-                    .create_new(true)
-                    .write(true)
-                    .open(&self.to)?;
-                to_file.set_permissions(metadata.permissions())?;
+                to_file.set_permissions(metadata.permissions()).await?;
+
+                // Prevent spamming the progress callbacks.
+                let mut last_progress_update = Instant::now();
+                // io_uring/IOCP requires transferring ownership of the buffer to the kernel.
+                let mut buf_in = std::mem::take(&mut ctx.buf);
+                // Track where the current read/write position is at.
+                let mut pos = 0;
+
                 loop {
-                    ctx.controller.check()?;
+                    let BufResult(result, buf_out) = from_file.read_at(buf_in, pos).await;
 
-                    let count = from_file.read(&mut ctx.buf)?;
-                    if count == 0 {
-                        break;
+                    let count = match result {
+                        Ok(0) => {
+                            ctx.buf = buf_out;
+                            break;
+                        }
+                        Ok(count) => count,
+                        Err(why) => {
+                            ctx.buf = buf_out;
+                            return Err(why.into());
+                        }
+                    };
+
+                    let BufResult(result, buf_out) = to_file.write_at(buf_out, pos).await;
```

Comment:
Is buf_out exactly count bytes?

Reply:
Fixed in https://github.com/pop-os/cosmic-files/commit/d6f1efbf672ed400362736550e838a3e6508e71d